src/catchup/CatchupManagerImpl.cpp

```diff
@@ -160,32 +179,32 @@ CatchupManagerImpl::processLedger(LedgerCloseData const& ledgerData)
         auto const& config = mCatchupWork->getCatchupConfiguration();
         if (ledgerData.getLedgerSeq() <= config.toLedger())
         {
-            return;
+            return false;
         }
 
-        addAndTrimSyncingLedgers(ledgerData);
+        trimSyncingLedgers();
         logAndUpdateCatchupStatus(true);
-        return;
+        return false;
     }
 
     // Next, we buffer every out of sync ledger to allow us to get back in sync
     // in case the ledgers we're missing are received.
-    addAndTrimSyncingLedgers(ledgerData);
+    trimSyncingLedgers();
 
     // Finally we wait some number of ledgers beyond the smallest buffered
     // checkpoint ledger before we trigger the CatchupWork. This could be any
     // number, for the sake of simplicity at the moment it's set to one ledger
     // after the first buffered one. Since we can receive out of order ledgers,
     // we just check for any ledger larger than the checkpoint
 
-    auto& hm = mApp.getHistoryManager();
-
     std::string message;
     uint32_t firstLedgerInBuffer = mSyncingLedgers.begin()->first;
     uint32_t lastLedgerInBuffer = mSyncingLedgers.crbegin()->first;
     if (mApp.getConfig().modeDoesCatchupWithBucketList() &&
-        hm.isFirstLedgerInCheckpoint(firstLedgerInBuffer) &&
-        firstLedgerInBuffer < lastLedgerInBuffer)
+        HistoryManager::isFirstLedgerInCheckpoint(firstLedgerInBuffer,
+                                                  mApp.getConfig()) &&
+        firstLedgerInBuffer < lastLedgerInBuffer &&
+        !mApp.getLedgerManager().isApplying())
```

Comment:
`isApplying` indicates that background thread currently has ledgers to apply. We shouldn't start catchup in this case, because it may conflict with background application. So what happens is we first make sure we've applied all queued ledgers, and only then go into catchup. If catchup is running, we should never apply any buffered ledgers (this is the same as master today).